File: Frontend/TypingPass/src/dev/vale/typing/Hinputs.scala
package dev.vale.typing

import dev.vale.postparsing.{IRuneS, ITemplataType}
import dev.vale.typing.ast.{EdgeT, FunctionDefinitionT, FunctionExportT, FunctionExternT, InterfaceEdgeBlueprintT, KindExportT, KindExternT, PrototypeT, SignatureT}
import dev.vale.typing.names.{CitizenNameT, CitizenTemplateNameT, FunctionNameT, IFunctionNameT, IdT, LambdaCitizenNameT}
import dev.vale.typing.templata._
import dev.vale.typing.types._
import dev.vale.{PackageCoordinate, StrI, vassert, vassertOne, vassertSome, vcurious, vfail, vimpl}
import dev.vale.typing.ast._
import dev.vale.typing.names._
import dev.vale.typing.types._

import scala.collection.mutable

case class InstantiationBoundArgumentsT(
  runeToFunctionBoundArg: Map[IRuneS, PrototypeT],
  runeToImplBoundArg: Map[IRuneS, IdT[IImplNameT]])

case class Hinputs(
  interfaces: Vector[InterfaceDefinitionT],
  structs: Vector[StructDefinitionT],
//  emptyPackStructRef: StructTT,
  functions: Vector[FunctionDefinitionT],
//  immKindToDestructor: Map[KindT, PrototypeT],

  // The typing pass keys this by placeholdered name, and the instantiator keys this by non-placeholdered names
  interfaceToEdgeBlueprints: Map[IdT[IInterfaceNameT], InterfaceEdgeBlueprintT],
  // The typing pass keys this by placeholdered name, and the instantiator keys this by non-placeholdered names
  interfaceToSubCitizenToEdge: Map[IdT[IInterfaceNameT], Map[IdT[ICitizenNameT], EdgeT]],

  instantiationNameToInstantiationBounds: Map[IdT[IInstantiationNameT], InstantiationBoundArgumentsT],

  kindExports: Vector[KindExportT],
  functionExports: Vector[FunctionExportT],
  kindExterns: Vector[KindExternT],
  functionExterns: Vector[FunctionExternT],
) {

  private val subCitizenToInterfaceToEdgeMutable = mutable.HashMap[IdT[ICitizenNameT], mutable.HashMap[IdT[IInterfaceNameT], EdgeT]]()
  interfaceToSubCitizenToEdge.foreach({ case (interface, subCitizenToEdge) =>
    subCitizenToEdge.foreach({ case (subCitizen, edge) =>
      subCitizenToInterfaceToEdgeMutable
        .getOrElseUpdate(subCitizen, mutable.HashMap[IdT[IInterfaceNameT], EdgeT]())
        .put(interface, edge)
    })
  })
  val subCitizenToInterfaceToEdge: Map[IdT[ICitizenNameT], Map[IdT[IInterfaceNameT], EdgeT]] =
    subCitizenToInterfaceToEdgeMutable.mapValues(_.toMap).toMap

  override def equals(obj: Any): Boolean = vcurious(); override def hashCode(): Int = vfail() // Would need a really good reason to hash something this big

  def lookupStruct(structId: IdT[IStructNameT]): StructDefinitionT = {
    vassertSome(structs.find(_.instantiatedCitizen.id == structId))
  }

  def lookupStructByTemplate(structTemplateName: IStructTemplateNameT): StructDefinitionT = {
    vassertSome(structs.find(_.instantiatedCitizen.id.localName.template == structTemplateName))
  }

  def lookupInterfaceByTemplate(interfaceTemplateName: IInterfaceTemplateNameT): InterfaceDefinitionT = {
    vassertSome(interfaces.find(_.instantiatedCitizen.id.localName.template == interfaceTemplateName))
  }

  def lookupImplByTemplate(implTemplateName: IImplTemplateNameT): EdgeT = {
    vassertSome(interfaceToSubCitizenToEdge.flatMap(_._2.values).find(_.edgeId.localName.template == implTemplateName))
  }

  def lookupInterface(interfaceId: IdT[IInterfaceNameT]): InterfaceDefinitionT = {
    vassertSome(interfaces.find(_.instantiatedCitizen.id == interfaceId))
  }

  def lookupEdge(implId: IdT[IImplNameT]): EdgeT = {
    vassertOne(interfaceToSubCitizenToEdge.flatMap(_._2.values).find(_.edgeId == implId))
  }

  def getInstantiationBoundArgs(instantiationName: IdT[IInstantiationNameT]): InstantiationBoundArgumentsT = {
    vassertSome(instantiationNameToInstantiationBounds.get(instantiationName))
  }

  def lookupStructByTemplateId(structTemplateId: IdT[IStructTemplateNameT]): StructDefinitionT = {
    vassertSome(structs.find(_.templateName == structTemplateId))
  }

  def lookupInterfaceByTemplateId(interfaceTemplateId: IdT[IInterfaceTemplateNameT]): InterfaceDefinitionT = {
    vassertSome(interfaces.find(_.templateName == interfaceTemplateId))
  }

  def lookupCitizenByTemplateId(interfaceTemplateId: IdT[ICitizenTemplateNameT]): CitizenDefinitionT = {
    interfaceTemplateId match {
      case IdT(packageCoord, initSteps, t: IStructTemplateNameT) => {
        lookupStructByTemplateId(IdT(packageCoord, initSteps, t))
      }
      case IdT(packageCoord, initSteps, t: IInterfaceTemplateNameT) => {
        lookupInterfaceByTemplateId(IdT(packageCoord, initSteps, t))
      }
    }
  }

  def lookupStructByTemplateName(structTemplateName: StructTemplateNameT): StructDefinitionT = {
    vassertOne(structs.filter(_.templateName.localName == structTemplateName))
  }

  def lookupInterfaceByTemplateName(interfaceTemplateName: InterfaceTemplateNameT): InterfaceDefinitionT = {
    vassertSome(interfaces.find(_.templateName.localName == interfaceTemplateName))
  }

  def lookupFunction(signature2: SignatureT): Option[FunctionDefinitionT] = {
    functions.find(_.header.toSignature == signature2).headOption
  }

  def lookupFunction(funcTemplateName: IFunctionTemplateNameT): Option[FunctionDefinitionT] = {
    functions.find(_.header.id.localName.template == funcTemplateName).headOption
  }

  def lookupFunction(humanName: String): FunctionDefinitionT = {
    val matches = functions.filter(f => {
      f.header.id.localName match {
        case FunctionNameT(n, _, _) if n.humanName.str == humanName => true
        case _ => false
      }
    })
    if (matches.size == 0) {
      vfail("Function \"" + humanName + "\" not found!")
    } else if (matches.size > 1) {
      vfail("Multiple found!")
    }
    matches.head
  }

  def lookupStruct(humanName: String): StructDefinitionT = {
    val matches = structs.filter(s => {
      s.templateName.localName match {
        case StructTemplateNameT(n) if n.str == humanName => true
        case _ => false
      }
    })
    if (matches.size == 0) {
      vfail("Struct \"" + humanName + "\" not found!")
    } else if (matches.size > 1) {
      vfail("Multiple found!")
    }
    matches.head
  }

  def lookupImpl(
    subCitizenTT: IdT[ICitizenNameT],
    interfaceTT: IdT[IInterfaceNameT]):
  EdgeT = {
    vassertSome(
      vassertSome(interfaceToSubCitizenToEdge.get(interfaceTT))
        .get(subCitizenTT))
  }

  def lookupInterface(humanName: String): InterfaceDefinitionT = {
    val matches = interfaces.filter(s => {
      s.templateName.localName match {
        case InterfaceTemplateNameT(n) if n.str == humanName => true
        case _ => false
      }
    })
    if (matches.size == 0) {
      vfail("Interface \"" + humanName + "\" not found!")
    } else if (matches.size > 1) {
      vfail("Multiple found!")
    }
    matches.head
  }

  def lookupUserFunction(humanName: String): FunctionDefinitionT = {
    val matches =
      functions
        .filter(function => simpleName.unapply(function.header.id).contains(humanName))
        .filter(_.header.isUserFunction)
    if (matches.size == 0) {
      vfail("Not found!")
    } else if (matches.size > 1) {
      vfail("Multiple found!")
    }
    matches.head
  }

  def nameIsLambdaIn(name: IdT[IFunctionNameT], needleFunctionHumanName: String): Boolean = {
    val first = name.steps.head
    val lastTwo = name.steps.slice(name.steps.size - 2, name.steps.size)
    (first, lastTwo) match {
      case (
        FunctionNameT(FunctionTemplateNameT(StrI(hayFunctionHumanName), _), _, _),
        Vector(
          LambdaCitizenTemplateNameT(_),
          LambdaCallFunctionNameT(LambdaCallFunctionTemplateNameT(_, _), _, _)))
        if hayFunctionHumanName == needleFunctionHumanName => true
      case _ => false
    }
  }

  def lookupLambdasIn(needleFunctionHumanName: String): Vector[FunctionDefinitionT] = {
    functions.filter(f => nameIsLambdaIn(f.header.id, needleFunctionHumanName)).toVector
  }

  def lookupLambdaIn(needleFunctionHumanName: String): FunctionDefinitionT = {
    vassertOne(lookupLambdasIn(needleFunctionHumanName))
  }

  def getAllNonExternFunctions: Iterable[FunctionDefinitionT] = {
    functions.filter(!_.header.isExtern)
  }

  def getAllUserFunctions: Iterable[FunctionDefinitionT] = {
    functions.filter(_.header.isUserFunction)
  }
}

File: FrontendRust/src/instantiating/ast/hinputs.rs
// VISTODO: rename Hinputs everywhere
use crate::utils::arena_index_map::ArenaIndexMap;
use crate::postparsing::names::IRuneS;
use crate::instantiating::ast::types::{cI, sI, StructIT};
use crate::instantiating::ast::names::IdI;
use crate::instantiating::ast::names::{IStructTemplateNameI, IInterfaceTemplateNameI, IImplTemplateNameI, IFunctionTemplateNameI};
use crate::instantiating::ast::ast::{
    EdgeI, FunctionDefinitionI, FunctionExportI, FunctionExternI, InterfaceEdgeBlueprintI,
    KindExportI, KindExternI, PrototypeI,
};
use crate::instantiating::ast::citizens::{ICitizenDefinitionI, InterfaceDefinitionI, StructDefinitionI};
use crate::instantiating::ast::names::INameI;
use crate::instantiating::ast::names::StructNameI;

/// Temporary state (see @TFITCX)
pub struct InstantiationBoundArgumentsI<'s, 'i> where 's: 'i {
    pub rune_to_function_bound_arg: ArenaIndexMap<'i, IRuneS<'s>, &'i PrototypeI<'s, 'i, sI>>,
    pub caller_rune_to_callee_rune_to_reachable_func:
        ArenaIndexMap<'i, IRuneS<'s>, ArenaIndexMap<'i, IRuneS<'s>, &'i PrototypeI<'s, 'i, sI>>>,
    pub rune_to_impl_bound_arg: ArenaIndexMap<'i, IRuneS<'s>, IdI<'s, 'i, sI>>,
}

/// Temporary state (see @TFITCX) — top-level container for instantiated output.
pub struct HinputsI<'s, 'i> where 's: 'i {
    pub interfaces: &'i [InterfaceDefinitionI<'s, 'i, cI>],
    pub structs: &'i [&'i StructDefinitionI<'s, 'i, cI>],
    pub functions: &'i [&'i FunctionDefinitionI<'s, 'i>],
    pub interface_to_edge_blueprints:
        ArenaIndexMap<'i, IdI<'s, 'i, cI>, InterfaceEdgeBlueprintI<'s, 'i>>,
    pub interface_to_sub_citizen_to_edge:
        ArenaIndexMap<'i, IdI<'s, 'i, cI>, ArenaIndexMap<'i, IdI<'s, 'i, cI>, EdgeI<'s, 'i>>>,
    pub kind_exports: &'i [KindExportI<'s, 'i>],
    pub function_exports: &'i [FunctionExportI<'s, 'i>],
    pub kind_externs: ArenaIndexMap<'i, &'i StructIT<'s, 'i, cI>, KindExternI<'s, 'i>>,
    pub function_externs: &'i [FunctionExternI<'s, 'i>],
}

impl<'s, 'i> HinputsI<'s, 'i> where 's: 'i {
    pub fn to_string(&self) -> String {
        panic!("Unimplemented: to_string")
    }
    pub fn lookup_function_by_str(&self, human_name: &str) -> &'i FunctionDefinitionI<'s, 'i> {
        let matches: Vec<&&'i FunctionDefinitionI<'s, 'i>> = self.functions.iter().filter(|f| {
            match f.header.id.local_name {
                INameI::FunctionNameIX(n) => n.template.human_name.0 == human_name,
                _ => false,
            }
        }).collect();
        if matches.is_empty() {
            panic!("Function \"{}\" not found!", human_name);
        } else if matches.len() > 1 {
            panic!("Multiple found!");
        }
        matches[0]
    }

// (Realized by `impl PartialEq for HinputsI` below.)

// (Realized by `impl Hash for HinputsI` below.)

    pub fn lookup_struct(
        &self,
        _struct_id: &IdI<'s, 'i, cI>,
    ) -> &'i StructDefinitionI<'s, 'i, cI> {
        self.structs.iter().find(|s| s.instantiated_citizen.id == *_struct_id).copied().expect("lookup_struct: not found")
    }

    pub fn lookup_interface(
        &self,
        _interface_id: &IdI<'s, 'i, cI>,
    ) -> &'i InterfaceDefinitionI<'s, 'i, cI> {
        self.interfaces.iter().find(|i| i.instantiated_interface.id == *_interface_id).expect("lookup_interface: not found")
    }

    pub fn lookup_citizen(
        &self,
        _citizen_id: &IdI<'s, 'i, cI>,
    ) -> ICitizenDefinitionI<'s, 'i, cI> {
        panic!("Unimplemented: lookup_citizen")
    }

    pub fn lookup_struct_by_template(
        &self,
        _struct_template_name: &IStructTemplateNameI<'s, 'i, cI>,
    ) -> &'i StructDefinitionI<'s, 'i, cI> {
        panic!("Unimplemented: lookup_struct_by_template")
    }

    pub fn lookup_interface_by_template(
        &self,
        _interface_template_name: &IInterfaceTemplateNameI<'s, 'i, cI>,
    ) -> &'i InterfaceDefinitionI<'s, 'i, cI> {
        panic!("Unimplemented: lookup_interface_by_template")
    }

    pub fn lookup_impl_by_template(
        &self,
        _impl_template_name: &IImplTemplateNameI<'s, 'i, cI>,
    ) -> &'i EdgeI<'s, 'i> {
        panic!("Unimplemented: lookup_impl_by_template")
    }

    pub fn lookup_edge(
        &self,
        _impl_id: &IdI<'s, 'i, cI>,
    ) -> &'i EdgeI<'s, 'i> {
        panic!("Unimplemented: lookup_edge")
    }

    pub fn lookup_function_by_template(
        &self,
        _func_template_name: &IFunctionTemplateNameI<'s, 'i, cI>,
    ) -> Option<&'i FunctionDefinitionI<'s, 'i>> {
        panic!("Unimplemented: lookup_function_by_template")
    }

    pub fn lookup_function(
        &self,
        _human_name: &str,
    ) -> &'i FunctionDefinitionI<'s, 'i> {
        panic!("Unimplemented: lookup_function")
    }

    pub fn lookup_struct_by_name(
        &self,
        human_name: &str,
    ) -> &'i StructDefinitionI<'s, 'i, cI> {
        let matches: Vec<&&'i StructDefinitionI<'s, 'i, cI>> = self.structs.iter().filter(|s| {
            match s.instantiated_citizen.id.local_name {
                INameI::StructName(StructNameI { template: IStructTemplateNameI::StructTemplate(t), .. }) if t.human_name.0 == human_name => true,
                _ => false,
            }
        }).collect();
        if matches.is_empty() {
            panic!("Struct \"{}\" not found!", human_name);
        } else if matches.len() > 1 {
            panic!("Multiple found!");
        }
        matches[0]
    }

    pub fn lookup_impl(
        &self,
        _sub_citizen_it: &IdI<'s, 'i, cI>,
        _interface_it: &IdI<'s, 'i, cI>,
    ) -> &'i EdgeI<'s, 'i> {
        panic!("Unimplemented: lookup_impl")
    }

    pub fn lookup_interface_by_name(
        &self,
        _human_name: &str,
    ) -> &'i InterfaceDefinitionI<'s, 'i, cI> {
        panic!("Unimplemented: lookup_interface_by_name")
    }

    pub fn lookup_user_function(
        &self,
        _human_name: &str,
    ) -> &'i FunctionDefinitionI<'s, 'i> {
        panic!("Unimplemented: lookup_user_function")
    }

    pub fn get_all_non_extern_functions(
        &self,
    ) -> Vec<&'i FunctionDefinitionI<'s, 'i>> {
        panic!("Unimplemented: get_all_non_extern_functions")
    }

    pub fn get_all_user_functions(
        &self,
    ) -> Vec<&'i FunctionDefinitionI<'s, 'i>> {
        panic!("Unimplemented: get_all_user_functions")
    }
}
